### PR TITLE
[FLOC-2012] Handle unknown other node state somewhat better in P2PManifestationDeployer

### PR DESCRIPTION
**Pull request by itamarst**
Fixes: https://clusterhq.atlassian.net/browse/FLOC-2012

### CODE REVIEW

File: flocker/node/_deploy.py [hunk 1]
```diff
@@ -930,15 +932,11 @@ def find_dataset_changes(uuid, current_state, desired_state):
                     going.add(DatasetHandoff(
                         dataset=dataset, hostname=hostname))
 
-    # Look at each dataset that is going to be hosted on this node.  If it
-    # was running somewhere else, we want that dataset to be in `coming`.
-    coming_dataset_ids = local_desired_dataset_ids.intersection(
-        remote_current_dataset_ids)
-    coming = set(dataset for dataset in local_desired_datasets
-                 if dataset.dataset_id in coming_dataset_ids)
-
```
Comment (wallrj) [inline flocker/node/_deploy.py]: :question: This seems unrelated. Is it just house keeping? I guess so, since you didn't have to make any corresponding test modifications.

File: flocker/node/test/test_deploy.py [hunk 2]
```diff
@@ -2196,6 +2196,38 @@ def test_unknown_applications(self):
...
+        The fact that a different node is ignorant about its manifestations
+        does not prevent calculating changes necessary for the current
+        node.
+        """
+        node_state = NodeState(
+            hostname=u"10.1.1.1",
+            uuid=uuid4(),
+            manifestations={MANIFESTATION.dataset_id:
+                            MANIFESTATION},
+            devices={}, paths={},
```
Comment (wallrj) [inline flocker/node/test/test_deploy.py]: I didn't think this should be allowed...
```
# Dataset attributes that must all be non-None if one is non-None:
_DATASET_ATTRIBUTES = {"manifestations", "paths", "devices"}
```

File: flocker/node/test/test_deploy.py [hunk 3]
```diff
@@ -2196,6 +2196,38 @@ def test_unknown_applications(self):
...
+            Node(hostname=api.hostname, uuid=api.node_uuid,
+                 manifestations=node_state.manifestations.transform(
+                     (DATASET_ID, "dataset", "deleted"), True))])
+
+        changes = api.calculate_changes(desired, current)
+        expected = sequentially(changes=[
+            in_parallel(changes=[DeleteDataset(dataset=DATASET.set(
+                "deleted", True))])
+            ])
+        self.assertEqual(expected, changes)
```
Comment (wallrj) [inline flocker/node/test/test_deploy.py]: Any way to use `assert_calculated_changes` here too?

Comment (wallrj) [inline flocker/node/test/test_deploy.py]: Ah, I see it hard codes `BlockDeviceDeployer` so never mind.

**Comment by wallrj** [comment]
Thanks @itamarst 

I left a few comments and questions above but it all makes sense. Please merge when you're satisfied with the build results.

File: flocker/node/_deploy.py (hunk 1 above)
Comment (itamarst) [inline flocker/node/_deploy.py]: Mostly, yeah. Although originally I thought deleting that code would suffice, until I reached the creation code, and figured I may as well leave that improvement in.

File: flocker/node/test/test_deploy.py (hunk 2 above)
Comment (itamarst) [inline flocker/node/test/test_deploy.py]: ? none of them are `None`, I don't see what should break.

Comment (wallrj) [inline flocker/node/test/test_deploy.py]: Oh, right, empty dicts. I misread. But the invariant should be stricter about this, right? If there are manifestations there must be devices and paths, otherwise they're non-manifest datasets.

Comment (itamarst) [inline flocker/node/test/test_deploy.py]: Not necessarily. ZFS doesn't have devices. For now this is fine, I think, improving manifests will happen some other time.
